File: docs/reference/python-langchain/app/whatsapp_provider.py
"""whatsapp_provider.py — abstraction over the WhatsApp channel so flow.py doesn't
care whether messages arrive via Twilio, Meta Cloud API, or local console testing.

To go live: set WHATSAPP_PROVIDER=twilio or =meta in .env and fill in the matching
credentials.
"""
import os
from typing import Optional

import httpx
from fastapi import Request


class InboundMessage:
    def __init__(self, from_: str, text: str):
        self.from_ = from_
        self.text = text


async def _console_parse_inbound(request: Request) -> Optional[InboundMessage]:
    body = await request.json()
    return InboundMessage(from_=body.get("from"), text=body.get("text"))


async def _console_send_message(to: str, text: str) -> None:
    print(f"\n[WhatsApp -> {to}]\n{text}\n")


async def _twilio_parse_inbound(request: Request) -> Optional[InboundMessage]:
    form = await request.form()
    from_ = (form.get("From") or "").replace("whatsapp:", "")
    return InboundMessage(from_=from_, text=form.get("Body"))


async def _twilio_send_message(to: str, text: str) -> None:
    from twilio.rest import Client

    client = Client(os.environ["TWILIO_ACCOUNT_SID"], os.environ["TWILIO_AUTH_TOKEN"])
    client.messages.create(
        from_=os.environ["TWILIO_WHATSAPP_FROM"],
        to=f"whatsapp:{to}",
        body=text,
    )


async def _meta_parse_inbound(request: Request) -> Optional[InboundMessage]:
    body = await request.json()
    entry = (body.get("entry") or [{}])[0]
    change = (entry.get("changes") or [{}])[0]
    message = (change.get("value", {}).get("messages") or [None])[0]
    if not message:
        return None
    return InboundMessage(from_=message.get("from"), text=message.get("text", {}).get("body", ""))


async def _meta_send_message(to: str, text: str) -> None:
    url = f"https://graph.facebook.com/v20.0/{os.environ['META_WA_PHONE_NUMBER_ID']}/messages"
    async with httpx.AsyncClient() as client:
        await client.post(
            url,
            headers={"Authorization": f"Bearer {os.environ['META_WA_ACCESS_TOKEN']}"},
            json={
                "messaging_product": "whatsapp",
                "to": to,
                "type": "text",
                "text": {"body": text},
            },
        )


_PROVIDERS = {
    "console": (_console_parse_inbound, _console_send_message),
    "twilio": (_twilio_parse_inbound, _twilio_send_message),
    "meta": (_meta_parse_inbound, _meta_send_message),
}


def get_provider():
    name = os.environ.get("WHATSAPP_PROVIDER", "console")
    parse_inbound, send_message = _PROVIDERS.get(name, _PROVIDERS["console"])
    return parse_inbound, send_message

File: docs/reference/agent-code/src/whatsappProvider.js
// whatsappProvider.js — abstraction over the WhatsApp channel so the flow logic
// doesn't care whether messages arrive via Twilio, Meta Cloud API, or local console testing.
//
// To go live: set WHATSAPP_PROVIDER=twilio or =meta in .env and fill in the matching
// credentials. Each provider must implement:
//   - parseInbound(req): (expressRequest) => { from: string, text: string }
//   - sendMessage(to, text): (string, string) => Promise<void>

const PROVIDER = process.env.WHATSAPP_PROVIDER || 'console';

const consoleProvider = {
  parseInbound(req) {
    // For local testing: POST { "from": "+919876543210", "text": "Hello Oscar" }
    return { from: req.body.from, text: req.body.text };
  },
  async sendMessage(to, text) {
    console.log(`\n[WhatsApp -> ${to}]\n${text}\n`);
  },
};

const twilioProvider = {
  parseInbound(req) {
    // Twilio webhook sends application/x-www-form-urlencoded with "From" and "Body"
    return { from: (req.body.From || '').replace('whatsapp:', ''), text: req.body.Body };
  },
  async sendMessage(to, text) {
    const twilio = require('twilio')(process.env.TWILIO_ACCOUNT_SID, process.env.TWILIO_AUTH_TOKEN);
    await twilio.messages.create({
      from: process.env.TWILIO_WHATSAPP_FROM,
      to: `whatsapp:${to}`,
      body: text,
    });
  },
};

const metaProvider = {
  parseInbound(req) {
    // Meta Cloud API webhook payload shape
    const entry = req.body.entry?.[0];
    const change = entry?.changes?.[0];
    const message = change?.value?.messages?.[0];
    if (!message) return null;
    return { from: message.from, text: message.text?.body || '' };
  },
  async sendMessage(to, text) {
    const url = `https://graph.facebook.com/v20.0/${process.env.META_WA_PHONE_NUMBER_ID}/messages`;
    await fetch(url, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${process.env.META_WA_ACCESS_TOKEN}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        messaging_product: 'whatsapp',
        to,
        type: 'text',
        text: { body: text },
      }),
    });
  },
};

const providers = { console: consoleProvider, twilio: twilioProvider, meta: metaProvider };

module.exports = providers[PROVIDER] || consoleProvider;
